Mini

latot opened the pull request:
Hi again, other hyper minor changes, in the second commit https://github.com/hrydgard/ppsspp/commit/7c63d7b445c629548506f65579012db24f55927b its because the expression, `a&b?c:d` can be interpreted like `(a&b)?c:d` or `a&(b?c:d)`, i don't know what is correct or  what was your intentions so i use the what g++ syntax, `(a&b)?c:d`.
well and i don't know if is right the first commit, https://github.com/hrydgard/ppsspp/commit/9d572a45acd533bcf8000c8a2660a9f846aac5f6

idea of this, coverity and cppcheck things...

hrydgard committed:
Add a bunch of new features and fixes to the README for 1.2.0

hrydgard committed:
More README 1.2.0 updates

hrydgard committed:
Add another fix

latot committed:
initialize cpudetect

little fix ini cpudetect

latot committed:
remove doubles breaks

unknownbrackets commented:
IMHO, the precedence of ternaries is pretty clear.  I hate when seeing `((x) ? (y) : (z))`.  To me it's like seeing someone who doesn't feel comfortable with math, and not really different from seeing `(((x + 1) - (9 / 3)) + 2)`.

Extra parenthesis should only be used when it clarifies meaning or similar.  I would disagre with changing `x = y + z * w` to `x = y + (z * w)` and I also disagree with changing `x = y ? z : w` to `x = (y) ? z : w`.  It's the same.

-[Unknown]

latot commented:
okis.

latot commented:
```
#  define send_code(s, c, tree) \
     { if (z_verbose>2) fprintf(stderr,"\ncd %3d ",(c)); \
       send_bits(s, tree[c].Code, tree[c].Len); }
#endif
```
c should be signed or unsigned?

unknownbrackets commented on Core/HLE/HLE.cpp:
```diff
@@ -626,7 +626,7 @@ size_t hleFormatLogArgs(char *message, size_t sz, const char *argmask) {
 
 		case 'i':
-			APPEND_FMT("%d", regval);
+			APPEND_FMT("%u", regval);
```
This needs to stay `%d`.

-[Unknown]

latot replied:
mm, but regval is a unsigned int, so ```APPEND_FMT("%d", unsigned(regval));```??

hrydgard replied:
regval may just as well represent a sign as an unsigned value, there's no way to know. %d is as good as any value, just cast to int for clarity. That is:

    APPEND_FMT("%d", (int)regval);

I think the rest of the changes to %u are also unnecessary, and if anything should be done, it's this same thing.

latot replied:
sorry last comment is wrong, its posible use
```APPEND_FMT("%d", signed(regval));```
insted of ```int```?
in this cases are mutliples types of data, i think can be good avoid change it

unknownbrackets replied:
Just to explain, registers in processors don't really have the C "rules".  For example, an "s8" on the PSP's processor is:

```
+--------------------------------+
+                        SIIIIIII+
+--------------------------------+
                         ^ sign bit
                          ^ integer bits
 ^ other bits that are ignored
```

In a way, they're always u32 values.  The "s8" rule just explains which portion of those bits are relevant, and what they mean.

Here, we're taking a register value (`regval`), and using `'i'` to indicate "this register which contains 32 bits of data is to be understood as an s32".  There are other cases (other than `'i'`), where it's treated differently.

As far as signed/int, this code you're changing is actually adhering to an ABI.  It's IMHO less clear to use `signed` when we're certain `regval` will always be a `u32`.  If this code were to be reused for a 64-bit MIPS emulator (like the PS2) it would need significant changes, anyway.

-[Unknown]

latot replied:
:O okis thx the explication.
a little question, in Core/MIPS/MIPSVFPUUtils.cpp line 152:
```
void ReadVector(float *rd, VectorSize size, int reg) {
```
affects if we use u32 instead float to rd?
i think remove this type of lines can help:
```u32 *rdu = (u32 *)rd;```
convert float to u32 don't is a portable code...

unknownbrackets replied:
So, the VFPU in the PSP is a vector floating point processor unit.  Its registers are floats, which happen to also be 32 bits.

Some instructions (not many) treat these registers as integers (e.g. u32 or s32 or even packed other things), and in some cases we interact with them as u32s to simulate floating point behavior.  But their "natural" type is single-precision float.

-[Unknown]

latot replied:
mm, okis, exist some method/option to avoid transform from float* to u32*?

unknownbrackets replied:
We're explicitly wanting to look at the bits stored in that float and work with them individually.  It might make sense to use a union instead, and in some places we do.... but I don't think `ReadVector` should necessarily read into something other than a float.

-[Unknown]

latot replied:
Hi, well searching, with yours idea of use union, seems to remove this type of line:
```u32 *rdu = (u32 *)rd;```
its necesary use ```ReadVector``` with ```FLoatBits``` like:
```void ReadVector(FloatBits *rd, VectorSize size, int reg)```
this is only because ```ReadVector``` use floats and u32.
what do you think?

latot replied:
here a working tree of your idea of use FloatBits(union):
https://github.com/latot/ppsspp/tree/pointer

unknownbrackets commented:
I'm not sure about some of these changes and haven't looked at it thoroughly.  I don't think we want to change zlib, we should probably just check if they've made changes.

-[Unknown]

latot committed:
match printf types v2

match printf types v2

little fix

minor fix, match printf type

latot commented:
okis, fixed and now don't touch zlib.

latot commented:
i don't know if this is right: https://github.com/latot/ppsspp/commit/287bb275cee4a18110dd0aac4203230592c42639

should match with cpudetect initialization.

unknownbrackets committed:
Correct recovery of locked disk cache files.

We were keeping the index when creating a new file, rather than clearing
it.  Additionally, flush before all reads - in case of overlap.

unknownbrackets committed:
Handle floats and sign extension in HLE logs.

In some cases, we were previously logging sign-extended error codes.  This
handles that better using type_traits.

unknownbrackets committed:
Improve depth accuracy a bit in 16-bit slice mode.

unknownbrackets committed:
Switch to old-style depth in OpenGL.

Some devices really aren't handling the new method well.  There seem to be
accuracy problems when specifying the values to OpenGL.

unknownbrackets committed:
Bring back the depth range hack.

unknownbrackets committed:
Fallback to per-vertex depth rounding on GLES2.

Instead of not-working 24->16 now that it's disabled.

unknownbrackets committed:
Limit savestate screenshots to top left of screen.

Sometimes framebuffers have extra stuff outside, we don't need that in the
screenshot.  Fixes #8552.

hrydgard committed:
More features in 1.2.0. Split out pre-1.0.0 history to history.md

hrydgard committed:
Minor correction to README.md

hrydgard committed:
Update lang submodule

unknownbrackets committed:
Shutdown EGL properly when quitting render loop.

unknownbrackets committed:
Android: Restart the render loop on wake.

unknownbrackets committed:
UI: Keep focus on game after displaying details.

Fixes #8539.

unknownbrackets committed:
Fix debugging and depth packing with old depth.

Which means, OpenGL in general right now.

latot commented:
this need to be checked, my intentions its don't up any change except if you say or need something

unknownbrackets commented on ext/native/ext/vjson/main.cpp:
```diff
@@ -25,7 +25,7 @@ void populate_sources(const char *filter, std::vector<std::vector<char> > &sourc
 	}
 
-	printf("Loaded %d json files\n", sources.size());
+	printf("Loaded %d json files\n", (int)sources.size());
```
These changes look like bugfixes, albeit to code we don't actually use.  It looks like upstream was already changed differently:

https://github.com/vivkin/vjson/blob/master/main.cpp#L32

-[Unknown]

latot replied:
okidoki

unknownbrackets commented on ext/native/net/http_client.cpp:
```diff
@@ -126,7 +126,7 @@ void DeChunk(Buffer *inbuffer, Buffer *outbuffer, int contentLength, float *prog
 			return;
 		int chunkSize;
-		sscanf(line.c_str(), "%x", &chunkSize);
+		sscanf(line.c_str(), "%x", (unsigned int *)&chunkSize);
```
I think a cast of a pointer argument to `sscanf()` should almost be an error.  In this case, maybe `dechunkedBytes` and `chunkSize` would just both be `uint32_t`s?

-[Unknown]

unknownbrackets commented on Core/Loaders.cpp:
```diff
@@ -36,7 +36,7 @@
 
 FileLoader *ConstructFileLoader(const std::string &filename) {
-	if (filename.find("http://") == 0 || filename.find("https://") == 0)
+	if (filename.compare(0, 7, "http://") == 0 || filename.compare(0, 8, "https://") == 0)
```
I don't like adding magic numbers for no practical gain.  It just means someone will copy and paste another protocol in some day and it won't work because they forgot to change the `7` that they didn't realize had to be the length of the component.

I also don't think `filename.compare(0, strlen("http://"), "http://")` is worth doing, because I'm not a member of the We Enjoy Typing brigade.

I think we have a `startsWith` utility function somewhere if you feel like it really needs to change.

-[Unknown]

latot committed:
back, chenged in upstream

back, chenged in upstream

hrydgard commented:
This pull request seems to have broken and needs some sort of rebase or similar. Not sure about the value of much of it anyway, though some of the casts do make sense.

Anyway, closing for now.

latot commented:
mm, you are right, with the merge this is all mixed and is very hard to read the code...

I think is necessary fix types in scanf and similars, because when don't match the doc say the behavior is undefined.., i don't like that..., anyway, my first commits say what i was trying to do with this, do you recommend split this?, i think to fix the types you or @unknownbrackets can have better ideas to handle it, i only can send little changes.

Thx.!
